src/linear_algebra/invariant_submodule.lean

```diff
@@ -0,0 +1,140 @@
+  U.invariant_under T ↔ T '' U ⊆ U := by rw [← set.maps_to', U.invariant_under_iff_maps_to]
+
+variables (U V : submodule R E) (hUV : is_compl U V) (T : E →ₗ[R] E)
+
+local notation `pᵤ` := linear_proj_of_is_compl U V hUV
+local notation `pᵥ` := linear_proj_of_is_compl V U hUV.symm
+
+lemma invariant_under.linear_proj_comp_self_comp_linear_proj_eq
+  (h : U.invariant_under T) (x : E) : ↑(pᵤ (T (pᵤ x))) = T (pᵤ x) :=
+begin
+  rw linear_proj_of_is_compl_eq_self_iff,
+  exact h (coe_mem _),
+end
+
+lemma invariant_under_of_linear_proj_comp_self_comp_linear_proj_eq
+  (h : ∀ x : E, ↑(pᵤ (T (pᵤ x))) = T (pᵤ x)) : U.invariant_under T :=
+begin
+  intros u hu,
+  rw [mem_comap, ← linear_proj_of_is_compl_eq_self_iff hUV _,
+      ← (linear_proj_of_is_compl_eq_self_iff hUV u).mpr hu, h],
+end
+
+/-- `U` is invariant under `T` if and only if `pᵤ.comp (T.comp pᵤ) = T.comp pᵤ`,
+where `pᵤ` denotes the linear projection to `U` along `V` -/
+lemma invariant_under_iff_linear_proj_comp_self_comp_linear_proj_eq :
+  U.invariant_under T ↔ (∀ x : E, (pᵤ (T (pᵤ x)) : E) = T (pᵤ x)) :=
+⟨invariant_under.linear_proj_comp_self_comp_linear_proj_eq U V hUV T,
+ invariant_under_of_linear_proj_comp_self_comp_linear_proj_eq U V hUV T⟩
+
+/-- `V` is invariant under `T` if and only if `pᵤ.comp (T.comp pᵤ) = pᵤ.comp T`,
+where `pᵤ` denotes the linear projection to `U` along `V` -/
+lemma invariant_under'_iff_linear_proj_comp_self_comp_linear_proj_eq :
+  V.invariant_under T ↔ (∀ x : E, (pᵤ (T (pᵤ x)) : E) = pᵤ (T x)) :=
+by simp_rw [invariant_under_iff_linear_proj_comp_self_comp_linear_proj_eq _ _ hUV.symm,
+            linear_proj_of_is_compl_eq_self_sub_linear_proj, map_sub, sub_eq_self,
+            submodule.coe_sub, sub_eq_zero, eq_comm]
+
+/-- both `U` and `V` are invariant under `T` if and only if `T` commutes with `pᵤ`,
+where `pᵤ` denotes the linear projection to `U` along `V` -/
+lemma is_compl_invariant_under_iff_linear_proj_and_self_commute :
+  (U.invariant_under T ∧ V.invariant_under T) ↔ commute (U.subtype.comp pᵤ) T :=
+begin
+  simp_rw [commute, semiconj_by, linear_map.ext_iff, linear_map.mul_apply,
+           linear_map.comp_apply, U.subtype_apply],
+  split,
+  { rintros ⟨h1, h2⟩ x,
+    rw [← (U.invariant_under'_iff_linear_proj_comp_self_comp_linear_proj_eq V _ _).mp h2 x],
+    exact (linear_proj_of_is_compl_eq_self_iff hUV _).mpr
+      (h1 (set_like.coe_mem (pᵤ x))) },
+  { intros h,
+    split,
+    { simp_rw [U.invariant_under_iff_linear_proj_comp_self_comp_linear_proj_eq _ hUV, h,
+               linear_proj_of_is_compl_idempotent hUV],
+      exact λ x, rfl },
+    { simp_rw [U.invariant_under'_iff_linear_proj_comp_self_comp_linear_proj_eq _ hUV, h,
+               linear_proj_of_is_compl_idempotent hUV],
+      exact λ x, rfl } }
+end
+
+/-- `U` is invariant under the inverse of `T` if and only if `U ⊆ T(U)` -/
+lemma invariant_under_inverse_iff_subset_image [invertible T] :
+  U.invariant_under (⅟ T) ↔ ↑U ⊆ T '' U :=
```

Comment:
Could you also add the `equiv` version? I'm completely fine with keeping this, but since you reduce to the `equiv` case in the proof I'd say you may as well state it, I expect it to be useful.